src/librustc_infer/infer/combine.rs

```diff
@@ -649,27 +649,47 @@ impl TypeRelation<'tcx> for Generalizer<'_, 'tcx> {
     ) -> RelateResult<'tcx, &'tcx ty::Const<'tcx>> {
         assert_eq!(c, c2); // we are abusing TypeRelation here; both LHS and RHS ought to be ==
 
+        debug!("generalize: consts c={:?}", c);
         match c.val {
             ty::ConstKind::Infer(InferConst::Var(vid)) => {
-                let mut inner = self.infcx.inner.borrow_mut();
-                let variable_table = &mut inner.const_unification_table();
-                let var_value = variable_table.probe_value(vid);
+                let var_value =
+                    self.infcx.inner.borrow_mut().const_unification_table().probe_value(vid);
                 match var_value.val {
                     ConstVariableValue::Known { value: u } => self.relate(&u, &u),
                     ConstVariableValue::Unknown { universe } => {
                         if self.for_universe.can_name(universe) {
                             Ok(c)
                         } else {
-                            let new_var_id = variable_table.new_key(ConstVarValue {
-                                origin: var_value.origin,
-                                val: ConstVariableValue::Unknown { universe: self.for_universe },
-                            });
+                            let new_var_id =
+                                self.infcx.inner.borrow_mut().const_unification_table().new_key(
+                                    ConstVarValue {
+                                        origin: var_value.origin,
+                                        val: ConstVariableValue::Unknown {
+                                            universe: self.for_universe,
+                                        },
+                                    },
+                                );
                             Ok(self.tcx().mk_const_var(new_var_id, c.ty))
                         }
                     }
                 }
             }
-            ty::ConstKind::Unevaluated(..) if self.tcx().lazy_normalization() => Ok(c),
+            ty::ConstKind::Unevaluated(did, substs, promoted)
+                if self.tcx().lazy_normalization() =>
+            {
+                // We have to generalize inference variables used in the generic substitutions,
+                // as unevaluated consts may otherwise contain invalid inference variables.
+                let new_substs =
+                    self.relate_with_variance(ty::Variance::Invariant, &substs, &substs)?;
```

Comment:
Is this an existing pattern? It feels like a hack, because relating something to itself should be a no-op. Why does this work?

Reply:
This is an existing pattern, we already abuse type relations in a lot of places where want to traverse a given type/const/whatever:

https://github.com/rust-lang/rust/blob/e070765f6e6c1035c7c728be486e7e69cb2b4808/src/librustc_infer/infer/combine.rs#L534

or https://github.com/rust-lang/rust/blob/e070765f6e6c1035c7c728be486e7e69cb2b4808/src/librustc_infer/infer/nll_relate/mod.rs#L818

Reply:
Would you be able to give a small example of what's going wrong as part of a comment here?

Reply:
later this week :laughing: I have to think of a good example where this check is required and not just
a stopgap because we have unused substs in `ConstKind::Unevaluated`.